src/main/scala/mesosphere/marathon/core/volume/VolumesModule.scala

```diff
@@ -0,0 +1,75 @@
+package mesosphere.marathon.core.volume
+
+import com.wix.accord._
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.state.{ AppDefinition, Container }
+import mesosphere.marathon.state.{ PersistentVolume, Volume }
+import mesosphere.marathon.core.volume.impl._
+
+trait LocalVolumes {
+  self: VolumeProvider[PersistentVolume] =>
+
+  /** @return a stream of task local volumes, extrapolating them from the app spec */
+  def local(app: AppDefinition): Iterable[Task.LocalVolume] = {
+    self.apply(app.container).map{ volume => Task.LocalVolume(Task.LocalVolumeId(app.id, volume), volume) }
+  }
+  /** @return the aggregate mesos disk resources required for volumes */
+  def diskSize(container: Option[Container]): Double = {
+    self.apply(container).map(_.persistent.size).flatten.sum.toDouble
+  }
+}
+
+/**
+  * VolumeProvider is an interface implemented by storage volume providers
+  */
+trait VolumeProvider[+T <: Volume] {
+  /** name uniquely identifies this volume provider */
+  val name: String
+  /** validation implements a provider's volume validation rules */
+  val validation: Validator[Volume]
+  /** containerValidation implements a provider's container validation rules */
+  val containerValidation: Validator[Container]
+
+  /** apply scrapes volumes from an application definition that are supported this volume provider */
+  def apply(container: Option[Container]): Iterable[T]
```

Comment:
Am not a big fan of this signature:
- Options in negative positions, i.e. as arguments, are usually an anti-pattern
- `apply` does not really make sense here. Why not call it `collect`? Then it's clear what it does.